### PR TITLE
zlinux micro architecture detection

### PR DESCRIPTION
**Pull request by annaibm**
- zlinux version detection
- add tests 

resolves: https://github.com/adoptium/TKG/issues/471

### CODE REVIEW

File: src/org/openj9/envInfo/MachineInfo.java [hunk 1]
```diff
@@ -201,10 +203,32 @@ private void getSysInfo() {
...
+			microArchOutput= "IBM z13";
+		} 
+		// Check if the detected microarchitecture is supported
+		Set<String> supportedMicroArchitectures = new HashSet<>(Arrays.asList("skylake", "z15", "z14", "z13"));
+
+		if (supportedMicroArchitectures.stream().anyMatch(microArchOutput.toLowerCase()::contains)) {
+			putInfo(new Info("microArch", MICRO_ARCH_CMD, microArchOutput, null));
+		} else {
+			// The detected microarchitecture is not supported
+			System.out.println("Detected microarchitecture is not supported: " + microArchOutput);
```
Comment (llxia) [inline src/org/openj9/envInfo/MachineInfo.java:229]: This message `Detected microarchitecture is not supported` is misleading. We support more than just `"skylake", "z15", "z14", "z13"`. It also means that we have to maintain the `supportedMicroArchitectures` list separately. 

Please remove this block of code. Keep original code and add else to printout microArchOutput value.

**Review by llxia — CHANGES_REQUESTED** [review]

File: src/org/openj9/envInfo/MachineInfo.java (hunk 1 above)
Comment (annaibm) [reply src/org/openj9/envInfo/MachineInfo.java:229]: Should I remove the code in line 229 and print out microArchOutput value there instead?

File: src/org/openj9/envInfo/MachineInfo.java [hunk 2]
```diff
@@ -38,7 +38,9 @@ public class MachineInfo {
 	public static final String[] LINUX_OS_NAME_CMD = new String[] {"bash", "-c", "grep '^NAME' /etc/os-release | awk -F'=' ' gsub(/\"/,\"\") { print $2}'"};
 	public static final String[] LINUX_OS_VERSION_CMD = new String[] {"bash", "-c", "grep '^VERSION_ID' /etc/os-release | awk -F'=' ' gsub(/\"/,\"\") { print $2}'"};
 
-	public static final String[] MICRO_ARCH_CMD = new String[] {"bash", "-c", "cat /proc/cpuinfo | grep 'model name' | uniq"};
+	// public static final String[] MICRO_ARCH_CMD = new String[] {"bash", "-c", "cat /proc/cpuinfo | grep 'model name' | uniq"};
```
Comment (llxia) [inline src/org/openj9/envInfo/MachineInfo.java:41]: Please remove the comment code.

**Review by llxia — CHANGES_REQUESTED** [review]

File: examples/platformRequirements/playlist.xml [hunk 3]
```diff
@@ -37,8 +37,76 @@
...
+		</platformRequirementsList>
+	</test>
+
+	<!-- Test for z15.t01 microarchitecture -->
+	<test>
+		<testCaseName>test_microarchitecture_detection_z15_t01</testCaseName>
+		<command>echo "test microarchitecture detection for z15.t01"; \
+		$(TEST_STATUS)</command>
+		<platformRequirementsList>
+			<platformRequirements>z15.t01</platformRequirements>
```
Comment (llxia) [inline examples/platformRequirements/playlist.xml:56]: https://github.com/adoptium/TKG/pull/477#discussion_r1417843983 is not addressed. Please check the rest of PR

**Comment by llxia** [comment]
- Please link the issue https://github.com/adoptium/TKG/issues/471 in the commit message.
- we need to support `+`. (i.e., `<platformRequirements>z13+</platformRequirements>`)

File: scripts/testTKG/test_platformRequirements.py [hunk 4]
```diff
@@ -16,105 +16,134 @@
...
+				passed.add('test_microarchitecture_detection_z14')
+				passed.add('test_microarchitecture_detection_z13_and_above_0')
+			elif label_str[0] == "z13s":
+				passed.add('test_microarchitecture_detection_z13s_0')
+				passed.add('test_microarchitecture_detection_z13_and_above_0')
+			elif label_str[0] == "z13+":
+				passed.add('test_microarchitecture_detection_z13_and_above_0')
+			elif label_str[0] == "z13":
+				passed.add('test_microarchitecture_detection_z13_0')
+			else:
```
Comment (llxia) [inline scripts/testTKG/test_platformRequirements.py:138]: This code needs to be more robust. Please use numeric comparison. We also need to consider z14+, z15+, etc

if version >= 13, then it is z13+
if version >= 14, then it is z14+
...

File: scripts/testTKG/test_platformRequirements.py [hunk 5]
```diff
@@ -16,105 +16,134 @@
...
+		try:
+			if label_str[0] == "skylake":
+				passed.add('test_microarchitecture_detection_skylake_0')
+			elif label_str[0] == "z15":
+				passed.add('test_microarchitecture_detection_z15')
+				passed.add('test_microarchitecture_detection_z13_and_above_0')
+			elif label_str[0] == "z14":
+				passed.add('test_microarchitecture_detection_z14')
+				passed.add('test_microarchitecture_detection_z13_and_above_0')
+			elif label_str[0] == "z13s":
```
Comment (llxia) [inline scripts/testTKG/test_platformRequirements.py:131]: I am not aware that we have any test cases for `z13s`. I do not think we need this. It should belong to z13

File: examples/platformRequirements/playlist.xml [hunk 6]
```diff
@@ -37,8 +37,61 @@
...
+	<test>
+		<testCaseName>test_microarchitecture_detection_skylake</testCaseName>
+		<command>echo "test microarchitecture detection for skylake"; \
+		$(TEST_STATUS)</command>
+		<platformRequirementsList>
+			<platformRequirements>skylake</platformRequirements>
+		</platformRequirementsList>
+	</test>
 
+	<!-- Test for z15.t01 microarchitecture -->
```
Comment (llxia) [inline examples/platformRequirements/playlist.xml:50]: Misleading comment. The test is for z15, not z15.t01

File: src/org/openj9/envInfo/MachineInfo.java [hunk 7]
```diff
@@ -201,10 +202,31 @@ private void getSysInfo() {
...
+		} else if (microArchOutput.contains("8561")) {
+			microArch = "IBM z15 T01";
+		} else if (microArchOutput.contains("3907")) {
+			microArch = "IBM z14 ZR1";
+		} else if (microArchOutput.contains("3906")) {
+			microArch = "IBM z14";
+		} else if (microArchOutput.contains("2965")) {
+			microArch = "IBM z13s";
+		} else if (microArchOutput.contains("2964")) {
+			microArch= "IBM z13";
```
Comment (llxia) [inline src/org/openj9/envInfo/MachineInfo.java:220]: We do not need to know the sub-processor types. That means, z15 T02 and z15 T01 are z15. z14 ZR1 is z14. z13s is z13.

**Comment by annaibm** [comment]
run tkg-test

**Comment by annaibm** [comment]
run tkg-test

**Comment by annaibm** [comment]
run tkg-test

**Comment by annaibm** [comment]
run tkg-test

**Comment by annaibm** [comment]
run tkg-test

**Comment by annaibm** [comment]
run tkg-test

**Comment by smlambert** [comment]
I believe we will need to update the permissions of the Github actions in this repository in order to have them run properly (in this case triggered by a comment), much like the other repositories are being updated (example, much like the labeler change in this PR https://github.com/adoptium/ci-jenkins-pipelines/pull/954).

**Comment by annaibm** [comment]
run tkg-test

File: .github/workflows/testTKG.yml [hunk 8]
```diff
@@ -2,6 +2,10 @@ name: "Run TKG Tests"
 on:
   issue_comment:
     types: [created]
+permissions:
+    contents: write
+    pull-requests: write
+    issues: write
 jobs:
```
Comment (smlambert) [inline .github/workflows/testTKG.yml:9]: I think we will have to do this in a separate PR, merge it and then you will get to resume the testing you want to do.

**Comment by annaibm** [comment]
run tkg-test

**Comment by annaibm** [comment]
run tkg-test

**Comment by llxia** [comment]
@annaibm please create a separate issue with TKG PR build. For z versions, similar to https://github.com/adoptium/TKG/blob/0c5e49d22d17f1b91e0cc054b6b8cd047497d1ac/examples/platformRequirements/playlist.xml#L100, we should have something like `arch.390.z14` or `arch.390.z13+`. Not `microarch.z14`

**Comment by annaibm** [comment]
run tkg-test

**Comment by annaibm** [comment]
run tkg-test

**Comment by annaibm** [comment]
run tkg-test

**Comment by annaibm** [comment]
run tkg-test

**Comment by annaibm** [comment]
Grinder links:

File: src/org/testKitGen/TestInfoParser.java [hunk 9]
```diff
@@ -444,28 +444,42 @@ private boolean matchPlat(String fullSpec, String pr) {
 		}
 
 		if (prSplitOnDot[0].equals("arch") && (prSplitOnDot.length == 3)) {
-			String microArch = prSplitOnDot[2];
-			if (!microArch.equals(arg.getMicroArch())) {
-				return false;
+			String microArchVersion = prSplitOnDot[2];
+			String environmentMicroArch = arg.getMicroArch();
+			if (microArchVersion.endsWith("+")) {
+				int microArchVersionInt, environmentMicroArchVerInt;
```
Comment (llxia) [inline src/org/testKitGen/TestInfoParser.java:450]: Please initialize the values

File: src/org/testKitGen/TestInfoParser.java [hunk 10]
```diff
@@ -444,28 +444,42 @@ private boolean matchPlat(String fullSpec, String pr) {
...
+				try {
+					microArchVersionInt = extractVersionNumber(microArchVersion.substring(0, microArchVersion.length() - 1), "platformRequirement");
+					environmentMicroArchVerInt = extractVersionNumber(environmentMicroArch, "environmentMicroArch");
+				} catch (NumberFormatException e) {
+					System.err.println(e.getMessage());
+					return false;
+				}
+				return microArchVersionInt <= environmentMicroArchVerInt;
+			} else {
+				return environmentMicroArch.equals(microArchVersion);
```
Comment (llxia) [inline src/org/testKitGen/TestInfoParser.java:460]: - Please use https://github.com/adoptium/TKG/pull/477/files#diff-1ce05e4d3d428c3c8430b54f2905415f30fd13919ff8269e99c127485d69093eR462-R499 as reference. For number comparison, please make a function if needed.
- We need to compare the first character as well. i.e., for arch.390.z15+, we need to ensure `z` matches.

**Review by llxia — CHANGES_REQUESTED** [review]

**Review by llxia — CHANGES_REQUESTED** [review]

**Comment by llxia** [comment]
Thanks @annaibm . Please rerun tests with the latest change. And we should also test `os` cases.

**Comment by annaibm** [comment]
Grinder links:
1.test_arch_390_z14: https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/39815/ passed for z14 m/c
2. test_arch_390_z14: https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/39816/skipped when z15 m/c is used
3. test_not_arch_390: https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/39817/ skipped for s390x_linux
4. test_not_arch_390:https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/39819/- passed for aarch64_linux
5. TESTLIST=test_os_linux_ubuntu22,test_arch_390_z13,test_arch_390_z13plus,test_arch_390_z14,test_arch_390_z14plus,test_arch_390_z15,test_arch_390_z15plus,test_arch_x86_skylake:
https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/39814/

OS testcases:
1. TESTLIST=test_os_linux_ubuntu22,test_os_osx,test_osx_x86-64,test_not_linux_arch_x86,test_osx_x86-64_win_x86_aix_ppc64,test_os_linux_ubuntu20plus,test_os_linux_ubuntu20plus_rhel8plus,test_not_os_linux_ubuntu20plus: https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/39810/
2. TESTLIST=test_os_linux_ubuntu22,test_os_osx,test_osx_x86-64,test_not_linux_arch_x86,test_osx_x86-64_win_x86_aix_ppc64,test_os_linux_ubuntu20plus,test_os_linux_ubuntu20plus_rhel8plus,test_not_os_linux_ubuntu20plus: (ubu22s390x-rt-1.hursley.ibm.com)
https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/39807/
3. test_os_linux_ubuntu22 : https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/39823/ ([rhel8s390x-rt-2.hursley.ibm.com]
4. test_os_linux_ubuntu22 : https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/39826/ (ubu22s390x-rt-1.hursley.ibm.com)

**Comment by llxia** [comment]
> test_os_linux_ubuntu22 : https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/39826/ (ubu22s390x-rt-1.hursley.ibm.com)

can you rerun this build?

**Comment by annaibm** [comment]
test_os_linux_ubuntu22 : https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/39872/(ubu22s390x-rt-1.hursley.ibm.com)

**Review by llxia — APPROVED** [review]
LGTM

**Review by LongyuZhang — APPROVED** [review]
LGTM
